### ISSUE TITLE
Serve command not passing webpack `devServer` config to webpack-dev-server

### ISSUE THREAD

**Issue by lukereative** [webpack/webpack-cli#1469]
**Describe the bug**

Serve command not passing webpack `devServer` config to webpack-dev-server

#### webpack.dev.js
```
  devServer: {
    https: isUsingHTTPS && secureLocalDomain,
    host: "myreactapp.test",
    hot: true,
    port: 8080,
    publicPath: "/myreactapp",
    quiet: true,
    proxy: {
      '**': {
        target: localURL,
        secure: false,
        headers: {
          'X-Dev-Server-Proxy': localURL
        }
      }
    }
  }
```

#### command used
`webpack serve --config webpack.dev.js`

**What is the current behavior?**

The webpack server runs but doesn't pick up all the options from the `devServer` config. You can see this by `webpack output` being undefined in the CLI output, but also the proxy was also not working.

```
ℹ ｢wds｣: Project is running at http://myreactapp.test:8080/
ℹ ｢wds｣: webpack output is served from undefined
ℹ ｢wds｣: Content not from webpack is served from /path/to/my/reactapp
```

**To Reproduce**

#### Relevant entries from `package.json`
```json
"@webpack-cli/serve": "^1.0.1-alpha.5",
"webpack": "^5.0.0-beta.13",
"webpack-cli": "^4.0.0-beta.8",
```

1. Define a `devServer` config in your webpack config file
2. Run `webpack serve --config webpack.dev.js`

**Expected behavior**

Previously the dev server would pick up `devServer` config object from your webpack config file 

**Additional context**

If I change [line 28](https://github.com/webpack/webpack-cli/blob/1c8d0ce94c1ab55f2b0ec99f4c84ca677c0fc24d/packages/serve/src/startDevServer.ts#L28) from `startDevServer.ts` it works as expected but obviously this ignores the cli overrides.

```diff
- const server = new Server(compiler, options);
+ const server = new Server(compiler, devServerOptions);
```

CLI output
```
ℹ ｢wds｣: Project is running at http://myreactapp.test:8080/
ℹ ｢wds｣: webpack output is served from /eccreactapp/dist
ℹ ｢wds｣: Content not from webpack is served from /path/to/my/reactapp
```

**Comment by anshumanv** [comment]
Taking a look, thanks for the report!

**Comment by anshumanv** [comment]
Found the issue, will fix in a bit. 👍

**Comment by anshumanv** [comment]
Issue - devServer config was never passed fully to the server.

**Comment by anshumanv** [comment]
@evilebottnawi even after passing `clientLogLevel: "none"` or `silent` in server config, I'm still seeing logs, this should not be the case right? 

![image](https://user-images.githubusercontent.com/21009455/79414317-f1ca2e00-7fc7-11ea-9533-21a6829c7395.png)

**Comment by evilebottnawi** [comment]
@anshumanv yep, because it is logs from server, not from client

**Comment by anshumanv** [comment]
> @anshumanv yep, because it is logs from server, not from client

Yes, But it should pass the devServer object from the config right to DevServer right?

**Comment by evilebottnawi** [comment]
@anshumanv we should get `devServer` config, override specified CLI flags and send it to dev server

**Comment by anshumanv** [comment]
@evilebottnawi yep, fix is here - https://github.com/webpack/webpack-cli/pull/1470

**Comment by lukereative** [comment]
Fix is now here #1649

**Comment by gabriel-peracio** [comment]
This is a pretty major blocker for my workflow, hope this gets solved soon!

**Comment by ylemkimon** [comment]
#1649 is now merged.

**Comment by Johannestegner** [comment]
Will this be released in a new alpha/beta release soon? Would be wonderful to have it working :)

**Comment by gabriel-peracio** [comment]
I guess we need to wait for #1634?

Kinda weird that #1649 (the PR that fixes this issue) was merged, but has not been part of a release yet. Without #1649, devServer is basically broken (In my use case I need the proxy options, I have resorted to patching webpack)

**Comment by Johannestegner** [comment]
> 
> (In my use case I need the proxy options, I have resorted to patching webpack)

Ya, kinda been considering doing that myself! hehe :)

**Comment by gabriel-peracio** [comment]
Looks like this was fixed in [webpack-cli v4.0.0-beta.9](https://github.com/webpack/webpack-cli/tree/v4.0.0-beta.9)

I can't test it now, but if someone can confirm, we might be able to close this issue. If no one is available, I'll try to find some time to test it.

**Comment by Johannestegner** [comment]
Tried with latest, sadly, in my case, all webpack (and cli commands) have stopped working:

```
> cross-env NODE_ENV=development webpack serve

CLI for webpack must be installed.
  webpack-cli (https://github.com/webpack/webpack-cli)

We will use "npm" to install the CLI via "npm install -D".
Do you want to install 'webpack-cli' (yes/no): no
You need to install 'webpack-cli' to use webpack via CLI.
You can also install the CLI manually.
````

Following packages installed:

```
    "dependencies": {
        "@babel/runtime-corejs3": "^7.11.2"
    },
    "devDependencies": {
        "@babel/core": "^7.11.6",
        "@jitesoft/babel-preset-main": "^2.5.2",
        "@webpack-cli/serve": "^1.0.1-alpha.5",
        "babel-loader": "^8.1.0",
        "copy-webpack-plugin": "^6.1.1",
        "core-js": "^3.6.5",
        "cross-env": "^7.0.2",
        "css-loader": "^4.3.0",
        "file-loader": "^6.1.0",
        "imagemin-gifsicle": "^7.0.0",
        "imagemin-mozjpeg": "^9.0.0",
        "imagemin-optipng": "^8.0.0",
        "imagemin-svgo": "^8.0.0",
        "imagemin-webp-webpack-plugin": "^3.3.3",
        "imagemin-webpack": "^5.1.1",
        "mini-css-extract-plugin": "^0.11.2",
        "optimize-css-assets-webpack-plugin": "^5.0.4",
        "sass": "^1.26.11",
        "sass-loader": "^10.0.2",
        "style-loader": "^1.2.1",
        "terser-webpack-plugin": "^4.2.2",
        "url-loader": "^4.1.0",
        "webpack": "^5.0.0-rc.0",
        "webpack-cli": "^4.0.0-beta.9"
    }
```

Haven't had time to look in to it, (just saw that there where a new version out a few min ago and only have time to test quick before fixing food! hehe)...

**Comment by evilebottnawi** [comment]
You don't need `@webpack-cli/serve`, it is built-in feature

**Comment by Johannestegner** [comment]
> 
> 
> You don't need `@webpack-cli/serve`, it is built-in feature

Even without I still get the same errors :/  
Will check if I can figure out why when I'm done with food :)

**Comment by evilebottnawi** [comment]
Just note - replace optimize-css-assets-webpack-plugin on https://github.com/webpack-contrib/css-minimizer-webpack-plugin (there are bugs in architecture)

**Comment by Johannestegner** [comment]
Okay, after deleting node_modules and package files a couple of times and reinstalling and all that fun stuff I now get the following error:

```
> cross-env NODE_ENV=development webpack-cli serve

[webpack-cli] Promise rejection: TypeError: cli.commandLineArgs is not a function
[webpack-cli] TypeError: cli.commandLineArgs is not a function
    at Object.serve [as default] (D:\Repositories\js\testing-stuff\node_modules\@webpack-cli\serve\lib\index.js:25:31)
    at Function.run (D:\Repositories\js\testing-stuff\node_modules\webpack-cli\lib\commands\ExternalCommand.js:37:51)
    at Command.<anonymous> (D:\Repositories\js\testing-stuff\node_modules\webpack-cli\lib\utils\arg-parser.js:31:73)
    at Command.listener [as _actionHandler] (D:\Repositories\js\testing-stuff\node_modules\webpack-cli\node_modules\commander\index.js:426:31)
    at Command._parseCommand (D:\Repositories\js\testing-stuff\node_modules\webpack-cli\node_modules\commander\index.js:1002:14)
    at Command._dispatchSubcommand (D:\Repositories\js\testing-stuff\node_modules\webpack-cli\node_modules\commander\index.js:953:18)
    at Command._parseCommand (D:\Repositories\js\testing-stuff\node_modules\webpack-cli\node_modules\commander\index.js:970:12)
    at Command.parse (D:\Repositories\js\testing-stuff\node_modules\webpack-cli\node_modules\commander\index.js:801:10)
    at argParser (D:\Repositories\js\testing-stuff\node_modules\webpack-cli\lib\utils\arg-parser.js:132:27)
    at runCLI (D:\Repositories\js\testing-stuff\node_modules\webpack-cli\lib\bootstrap.js:25:24)
```

It does seem to be kind'a true, the `cli.commandLineArgs` property is _not_ a function, it actually does not exist in the `../lib/webpack-cli.js` file. I'm kinda wondering if something is wrong in the compilation of the script/s?

If you look at the [`index.ts`](https://github.com/webpack/webpack-cli/blob/next/packages/serve/src/index.ts) file in the serve package, it calls the parseArgs method, which is _not_ done in the actual package (which seems to be 1.0.1-alpha.5 when not including it in package file and allowing it to be installed with webpack-cli instead).

Changing to use `argParser` (which is what happens in the TS files) produces further errors.

Is the package installed with `webpack-cli` the correct package?

**Comment by anshumanv** [comment]
Yep this is known, https://github.com/webpack/webpack-cli/issues/1814

Releasing of sub packages should solve this, I think @evilebottnawi/@evenstensberg is already on this

**Comment by Johannestegner** [comment]
Aah, i see! That's good to hear, just built the latest serve version to test (from the next branch). Other than having to install `webpack-dev-server` (which I'm guessing is bundled in the 'real' installation) and comment out some of the configuration, it seems to run fine!

So I eagerly await the release of next serve version :)

**Comment by ylemkimon** [comment]
I can confirm that this issue is fixed with `webpack@5.0.0-rc.0` and `webpack-cli@4.0.0-rc.0`.

**Comment by evilebottnawi** [comment]
Let's close, feel free to open a new issue if you faced with the issue again
